recipes-extended/xen/files/libxl-blktap3-support.patch

```diff
@@ -0,0 +1,703 @@
++    char *fe_path = libxl__device_frontend_path(gc, aodev->dev);
+     char *state_path = GCSPRINTF("%s/state", be_path);
++    char *fe_state_path = GCSPRINTF("%s/state", fe_path);
+     char *online_path = GCSPRINTF("%s/online", be_path);
+     const char *state;
+     libxl_dominfo info;
+@@ -990,6 +987,11 @@ void libxl__initiate_device_generic_remo
+                 LOGD(ERROR, domid, "unable to write to xenstore path %s", state_path);
+                 goto out;
+             }
++            rc = libxl__xs_write_checked(gc, t, fe_state_path, GCSPRINTF("%d", XenbusStateClosed));
++            if (rc) {
++                LOGD(ERROR, domid, "unable to write to xenstore path %s", fe_state_path);
++                goto out;
++            }
+         }
+ 
+         rc = libxl__xs_transaction_commit(gc, &t);
+Index: xen-4.9.1/tools/libxl/libxl_disk.c
+===================================================================
+--- xen-4.9.1.orig/tools/libxl/libxl_disk.c
++++ xen-4.9.1/tools/libxl/libxl_disk.c
+@@ -56,7 +56,7 @@ static void disk_eject_xswatch_callback(
+             "/local/domain/%d/backend/%" TOSTRING(BACKEND_STRING_SIZE)
+            "[a-z]/%*d/%*d",
+            &disk->backend_domid, backend_type);
+-    if (!strcmp(backend_type, "tap") || !strcmp(backend_type, "vbd")) {
++    if (!strcmp(backend_type, "tap") || !strcmp(backend_type, "vbd") || !strcmp(backend_type, "vbd3")) {
+         disk->backend = LIBXL_DISK_BACKEND_TAP;
+     } else if (!strcmp(backend_type, "qdisk")) {
+         disk->backend = LIBXL_DISK_BACKEND_QDISK;
+@@ -189,7 +189,7 @@ int libxl__device_from_disk(libxl__gc *g
+             device->backend_kind = LIBXL__DEVICE_KIND_VBD;
+             break;
+         case LIBXL_DISK_BACKEND_TAP:
+-            device->backend_kind = LIBXL__DEVICE_KIND_VBD;
++            device->backend_kind = LIBXL__DEVICE_KIND_VBD3;
+             break;
+         case LIBXL_DISK_BACKEND_QDISK:
+             device->backend_kind = LIBXL__DEVICE_KIND_QDISK;
+@@ -310,7 +310,6 @@ static void device_disk_add(libxl__egc *
+             case LIBXL_DISK_BACKEND_PHY:
+                 dev = disk->pdev_path;
+ 
+-        do_backend_phy:
+                 flexarray_append(back, "params");
+                 flexarray_append(back, dev);
+ 
+@@ -322,27 +321,29 @@ static void device_disk_add(libxl__egc *
+                 break;
+ 
+             case LIBXL_DISK_BACKEND_TAP:
+-                if (dev == NULL) {
+-                    dev = libxl__blktap_devpath(gc, disk->pdev_path,
+-                                                disk->format);
+-                    if (!dev) {
+-                        LOGD(ERROR, domid, "Failed to get blktap devpath for %p",
+-                             disk->pdev_path);
+-                        rc = ERROR_FAIL;
+-                        goto out;
+-                    }
+-                }
+-                flexarray_append(back, "tapdisk-params");
++		rc = 0;
```

Comment:
this is pedantic

Reply:
I'm not sure what's pedantic here, could you please be more specific ?

Reply:
While the rc variable has been reused up to this point, every time it has been assigned a value, it is checked for a non-zero value and if true results in a jump to the out label. In your modifications you do not ever use rc (and thus not reliant on it actually being zero) except to assign it an error value, and the original author(s) already clear it for you at the end before going into the out labeled section. So IMHO setting rc to zero, when it is already expected to be zero, you not even dependent on it being zero, and your only operation to set the value to a non-zero value is a bit pedantic. For me, I am reviewing your patch in context of the function being modified and suddenly I see rc being zeroed then I have to stop and ask why did you do this, is there something here to be concerned with, if so, then why not fix up the rest of the function to provide the same assurances?

Reply:
rc being set to zero was not necessary, agreed. I will remove it but I didn't intentionally leave it there. There were issues interfacing with blktap3 because of which I had introduced it. After fix, I missed to remove it in my self review. Anyways, will fix it now.